### PR TITLE
Pipeline for streetview_v4 dataset for both Mixmatch and Resnet50

### PR DESCRIPTION
**Pull request by SwimmingPig**
This is a pipeline to generate texts using Google OCR on storefront images. Then the text is embedded into vectors using word2vec libraries.
Both the Mixmatch model and the Resnet50 model is updated for the new dataset format.

### CODE REVIEW

File: Mixmatch/libml/data.py [hunk 1]
```diff
@@ -35,16 +35,49 @@
...
+#     image = tf.image.decode_image(features['image'])
+#     image = tf.cast(image, tf.float32) * (2.0 / 255) - 1.0
+#     label = features['label']
+#     return dict(image=image, label=label)
+
+# Reshpae the embeddings to fit the size of a image channel.
+def reshape_embeddings(embeddings):
+    # EMBEDDING_MAX is the longest embedding size that can fit a channel, for image_size:64 and embeddings_size:300:
+    # the embedding_max is 64*64 // 300 * 300 = 13 * 300 = 3900
+    EMBEDDING_MAX = IMAGE_SIZE**2 // EMBEDDING_SIZE * EMBEDDING_SIZE
```
Comment (yuxizhang-google) [inline Mixmatch/libml/data.py]: nit: add space around **

File: Mixmatch/streetview_dataset/OCR_lib.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+def get_embeddings_from_image(img):
+    texts = detect_text(img)
+    
+    if texts:
+        embeddings = word_to_vec(texts).flatten()
+    else:
+        texts = " "
+        embeddings = np.array([0.0])
+    
+    return texts, embeddings
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/OCR_lib.py:86]: optional: you can also return a boolean here to indicate whether there are text detections. Then you can use it to analyze how many examples in the data set (especially the test data set) has texts and non-zero embeddings.

File: Supervised_learning/resnet_lib.py [hunk 3]
```diff
@@ -9,28 +10,69 @@
...
+    # EMBEDDING_MAX is the longest embedding size that can fit a channel, for image_size:64 and embeddings_size:300:
+    # the embedding_max is 64*64 // 300 * 300 = 13 * 300 = 3900
+    EMBEDDING_MAX = IMAGE_SIZE**2 // EMBEDDING_SIZE * EMBEDDING_SIZE
+    if tf.shape(embeddings) > EMBEDDING_MAX:
+        embeddings = tf.slice(embeddings, [0], [EMBEDDING_MAX])
+    zero_padding = tf.zeros([IMAGE_SIZE**2] - tf.shape(embeddings), dtype=tf.float32)
+    embeddings_padded = tf.concat([embeddings, zero_padding], 0)
+    result = tf.reshape(embeddings_padded, [IMAGE_SIZE, IMAGE_SIZE, 1])
+    
+    return result
```
Comment (yuxizhang-google) [inline Supervised_learning/resnet_lib.py]: Make a common library for this one and the reshape_embeddings in SSL.

File: Supervised_learning/resnet_lib.py [hunk 4]
```diff
@@ -9,28 +10,69 @@
...
     
     image, label = [], []
     
-    for image_features in parsed_image_dataset:
-        label.append(image_features['label'])
-        image.append(tf.image.decode_image(image_features['image']).numpy())
+    if word_embeddings:
+        for image_features in parsed_image_dataset:
+            img = tf.image.decode_image(image_features['image'])
+            img = tf.cast(img, tf.float32) * (2.0 / 255) - 1.0
```
Comment (yuxizhang-google) [inline Supervised_learning/resnet_lib.py:55]: Make a comment to explain this.

File: Mixmatch/streetview_dataset/README.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+| streetview_v3_256.1@250-200 |   256x256  |   250  |     0.0001    |      8     |   68.17  |
+|  streetview_v3_64.1@250-200 |    64x64   |   250  |     0.0001    |     64     |   76.66  |
+|  streetview_v3_64.1@250-200 |    64x64   |   250  |     0.0001    |      8     |   67.77  |
+
+
+### Streetview_v4:
+There are two ways to add the word embedding to the model, one way is to concatenate the vectors after the image pixels, the other way is to feed them into the model at the top fully connected layers.
+![Model Layer Depth to Accuracy](./IMG/streetview_v4_embedding.png?raw=true)
+
+| Label Size | Learning Rate | Batch Size | Accuracy w/o word embeddings | Accuracy w/ word embeddings A method | Accuracy w/ word embeddings B method |
```
Comment (yuxizhang-google) [inline Mixmatch/streetview_dataset/README.md]: Explain which approach (A or B) does this table use.

File: Supervised_learning/resnet_lib.py (hunk 3 above)
Comment (SwimmingPig) [reply Supervised_learning/resnet_lib.py]: Done. Move the function to OCR_lib.py, and both resnet_lib.py and data.py will import the function from there.

File: Supervised_learning/resnet_lib.py (hunk 4 above)
Comment (SwimmingPig) [reply Supervised_learning/resnet_lib.py:55]: Done

File: Mixmatch/streetview_dataset/README.md (hunk 5 above)
Comment (SwimmingPig) [reply Mixmatch/streetview_dataset/README.md]: Done

File: Supervised_learning/resnet_lib.py (hunk 4 above)
Comment (yuxizhang-google) [reply Supervised_learning/resnet_lib.py:55]: Sorry I didn't make it clear. I meant add comments to explain this formula:
`img = tf.cast(img, tf.float32) * (2.0 / 255) - 1.0`

Comment (SwimmingPig) [reply Supervised_learning/resnet_lib.py:55]: Added the comments.

**Review by yuxizhang-google — APPROVED** [review]
Great job!

**Pull request merged**